### PR TITLE
Expanded IFacade hook

### PR DESCRIPTION
**Pull request by CreativeMD**
Hey,

I finally found some time and tried to implement connected textures for LittleTiles (https://github.com/Chisel-Team/Chisel/issues/378). After reading some of your code, I spotted an interface which would allow me to implement exactly what i want: `IFacade`. Unfortunately, one parameter is missing, since there can be all kind of different block types inside one of my blocks and i don't know which one to return. So i added a `neighbor` parameter, which i can use to check if that block state can be found inside my block or not.

In Regards
CreativeMD

### CODE REVIEW

File: src/main/java/team/chisel/api/IFacade.java [hunk 1]
```diff
@@ -25,9 +25,11 @@
      *            The side being rendered, NOT the side being connected from.
      *            <p/>
      *            This value can be null if no side is specified. Please handle this appropriately.
+     * @param neighbor
+     *            The state to check against.
      * @return The blockstate which your block appears as.
      */
     @Nonnull
-    IBlockState getFacade(@Nonnull IBlockAccess world, @Nonnull BlockPos pos, @Nullable EnumFacing side);
+    IBlockState getFacade(@Nonnull IBlockAccess world, @Nonnull BlockPos pos, @Nullable EnumFacing side, IBlockState neighbor);
```
Comment (tterrag1098) [inline src/main/java/team/chisel/api/IFacade.java]: This param should be `@Nonnull`

**Review by tterrag1098 — CHANGES_REQUESTED** [review]
I think this change is ok, but I worry that it might still be too limited. For instance, what if you want to know more than state info about the block you're connecting to, such as TE or extended state data? This is going to be an API breaking change, so I want to make it worthwhile.

**Comment by CreativeMD** [comment]
I have not thought about it in that way. I don't plan to implement any calculations to check whether the two blocks connect directly or not. From my experience the required performance is not worth it and most of the times you want the textures to connect even if the tiles aren't placed exactly next to each other.

**Comment by HenryLoenwind** [comment]
If you add the new methods instead of replacing the old and then wrap the call in a try-catch block for ---whatsitcalled? AbstractMethodException?--- with the old one as fallback, this change will not be breaking. On the other hand, does anyone consume this API in 1.11 atm?

Then is the question on whether this change goes far enough. 

```
/**
Returns:
<li>null to force the other block to connect textures
<li>Blocks.AIR.getDefaultBlockstate() to force the other block to NOT connect textures
<li>the blockstate the block exposed at the given location, the caller decides on connecting or not
**/
IBlockState getFacade(@Nonnull IBlockAccess world, @Nonnull BlockPos pos, @Nullable FacadeSide side, @Nullable IBlockState neighborState)
...

enum FacadeSide {
DOWN, UP, EAST, WEST, SOUTH, NORTH, // the center of those faces
DOWN_EAST. DOWN_WEST, DOWN_SOUTH, DOWN_NORTH, // the 4 edges of the down side
UP_EAST. UP_WEST, UP_SOUTH, UP_NORTH, // the 4 edges of the up side
EAST_DOWN EAST_UP, EAST_SOUTH, EAST_NORTH, // the 4 edges of the up side
...
DOWN_EAST_NORTH, DOWN_EAST_SOUTH, DOWN_WEST_NORTH, DOWN_WEST_SOUTH, // the 4 corner of the down side,
...
// add methods to get the data from the names, getBaseSide, isEdge, isCorner, getEdgeDirection...

```

Edit: I'm not sure the corners are really needed. And the sides aren't needed for rendering, but for WAILA/TOP. Also, that enum could be replaced by 3 nullable EnumFacings.

An example: Ender IO has a painted double slab. So it would report the paint of the top half for the 4 edges of the top side and the top edges of the 4 sides around. It would do the respective thing for the bottom. And it would report AIR for the side edges of the side faces because they ar ehalf and half textured and connecting to either of those textures (or even both) would look weird.

**Comment by tterrag1098** [comment]
@HenryLoenwind the API you are describing there already exists. I was thinking something along the lines of 

```java
IBlockState getFacade(IBlockAccess world, BlockPos pos, EnumFacing side, Dir connection);
```

Where `Dir` is this class https://github.com/Chisel-Team/Chisel/blob/1.10/dev/src/main/java/team/chisel/common/util/Dir.java

The shortcoming of this is determining facade state for a non-ctm render, such as pillars, where the connecting state may not even be adjacent to the given block.

**Comment by HenryLoenwind** [comment]
That Dir is basically the same as the enum I posted, just without the base EnumFacing. But you have the headache of having to find out what the "left" side of the bottom face of a block is.

**Comment by tterrag1098** [comment]
There is already code to offset a blockpos, so that's not an issue

**Comment by CreativeMD** [comment]
Ok, i added a Dir parameter as well (not if this will work out though).

**Comment by tterrag1098** [comment]
The problem is, as mentioned above, the connection might not always be representable by a `Dir`. For instance in the case of pillars, they require knowledge about blocks which are 2 units away. So I think, instead of using an enum of any kind, just providing the second location (as a `BlockPos`) would be the best course of action. Perhaps providing a utility to turn that into a `Dir` if the API user would find that useful.

I think I will handle this change myself, as it will involve reworking some API classes.

**Comment by CreativeMD** [comment]
So i will close this pr. Just comment on the issue if there are any updates.

**Pull request closed without merging**